File: main/java/mod/iceandshadow3/compat/CRefEntity.java
package mod.iceandshadow3.compat;

import mod.iceandshadow3.basics.Damage;
import mod.iceandshadow3.basics.EDamageShape;
import mod.iceandshadow3.basics.EDamageType;
import net.minecraft.entity.Entity;
import net.minecraft.util.text.ITextComponent;

//TODO: Manually generated class stub.
public class CRefEntity implements IEffectSource {
	final Entity entity;
	
	CRefEntity(Entity e) {
		entity = e;
	}
	
	@Override
	public ITextComponent getNameTextComponent() {
		return entity.getDisplayName();
	}

	@Override
	public Entity getEffectSourceEntity() {
		return entity;
	}
	
	@Override
	public Damage getAttack() {
		if(entity instanceof IEffectSource) return ((IEffectSource)entity).getAttack();
		return null;
	}

	public void damage(Damage attack) {
		entity.attackEntityFrom(new ADamageSource(attack), attack.onDamage(this));
	}

}

File: main/scala/mod/iceandshadow3/compat/entity/CRefEntity.scala
package mod.iceandshadow3.compat.entity

import mod.iceandshadow3.basics.damage.Damage
import mod.iceandshadow3.compat.item.CRefItem
import mod.iceandshadow3.compat.world.TCWorldPlace
import mod.iceandshadow3.util.{EmptyIterator, IPositional, Vec3}
import net.minecraft.entity.Entity
import net.minecraft.util.text.ITextComponent

class CRefEntity(protected[compat] val entity: Entity) extends TCWorldPlace with TEffectSource with IPositional {
	override def getEffectSourceEntity: Entity = entity
	override def getNameTextComponent: ITextComponent = entity.getDisplayName
	override def position = new Vec3(entity.posX, entity.posY, entity.posZ)
	
	override protected[compat] def exposeWorld(): net.minecraft.world.World = entity.world

	override def getAttack: Damage = entity match {
		case source: TEffectSource => source.getAttack
		case _ => null
	}

	def damage(attack: Damage): Unit = 
		entity.attackEntityFrom(new ADamageSource(attack), attack.onDamage(this))

	def teleport(newpos: Vec3): Unit = {
		//TODO: For very long teleports, do we still need to do chunk loading shenanigans ala gatestones?
		val pitchyaw = entity.getPitchYaw
		entity.setPositionAndUpdate(newpos.xDouble, newpos.yDouble, newpos.zDouble)
	}

	def items(): Iterator[CRefItem] = new EmptyIterator[CRefItem]
}
